24297 - Alteration Business Type & Change Name Update

ArwenQin opened the pull request:
*Issue #:* /bcgov/entity#24297

*Description of changes:*
- update error showing logic - now show error message after user clicks "Done" button in Type Change section
- update error message format - in a message box, red outline
- When there is an error, mark this section invalid
- update wording and language in the error message
- now enable change a named business to numbered, and change its legal type in one filing (Case 7)
- fixed a bug in the "Cancel" button in type change section: it'a BC, 1 director -> enter the NR to change to a CC -> show error: less than 3 director-> click Cancel -> the name change and type change were incorrectly accepted
- fixed the error: change Named business to Numbered -> showing wrong numbered name, always showed "B.C. Ltd". 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

ArwenQin commented:
/gcbrun

bcregistry-sre commented:
Temporary Url for review: https://business-edit-dev--pr-600-w43kprmx.web.app

ArwenQin commented on src/components/common/YourCompany/ChangeBusinessType.vue:
```diff
@@ -191,7 +211,7 @@
             color="primary"
-            @click="isEditingType = false"
+            @click="resetType()"
           >
```
This is to fix a bug in the Cancel button:
 it'a BC, 1 director -> enter the NR to change to a CC -> show error: less than 3 director
![image](https://github.com/user-attachments/assets/a2f9efc8-a86e-4ac2-96e6-1184c4789c9b)

click Cancel -> the name change and type change were incorrectly accepted:
![image](https://github.com/user-attachments/assets/5dd7b200-38b6-4456-9fc8-86a20f8ee9e5)

severinbeauvais replied:
Good catch.

ArwenQin commented on src/components/common/YourCompany/CorrectName/CorrectNameToNumber.vue:
```diff
@@ -9,7 +9,7 @@
           class="mb-n5"
-          :label="`Change the company name to ${businessId} B.C. Ltd.`"
+          :label="`Change the company name to ${businessId} B.C. ${getUpdatedName()}`"
         />
```
fixed the error: change Named business to Numbered -> showing wrong numbered name, always showed "B.C. Ltd".

ArwenQin commented on src/components/common/YourCompany/EntityName.vue:
```diff
@@ -52,7 +52,8 @@
-              <span>The name of this business will be the current Incorporation Number followed by "B.C. Ltd."</span>
+              <span>The name of this business will be the current Incorporation Number followed
+                by "B.C. {{ getUpdatedName() }}"</span>
             </div>
```
fixed the error: change Named business to Numbered -> showing wrong numbered name, always showed "B.C. Ltd".

ArwenQin commented:
Case 7 - now enabled
This is a named business. -> User clicks Change Business Type first. -> selects a name-incompatible business type (eg, BC → CCC, BC → ULC, BEN → CCC, BEN → ULC, or vice-versa).
Result = error that name must be changed.
![image](https://github.com/user-attachments/assets/a0022e91-128e-48a6-8690-2eeacf69ab5c)

If user selects to use incorporation number as name then company becomes numbered (with compatible name), error goes away, and user needs to click business type checkbox and click Done.
![image](https://github.com/user-attachments/assets/90cdfa64-9de8-4cdf-872d-63eb63df66e6)

After clicking Done in Name Change Section:
![image](https://github.com/user-attachments/assets/407c8baa-3ea9-43ee-b416-a57c1544dc35)

Click Done in type change section:
![image](https://github.com/user-attachments/assets/72995b3c-5ae0-43e0-b6a6-3b94c673ab87)

Both numbered name and legal type are updated

ArwenQin commented on src/components/common/YourCompany/ChangeBusinessType.vue:
```diff
@@ -133,26 +133,47 @@
-          class="my-6"
+          class="mt-6"
+          color="red"
         >
```
Update the error message format and wording:
![image](https://github.com/user-attachments/assets/7138c679-49cb-4864-bd7b-d46539a24336)

severinbeauvais commented on src/components/common/YourCompany/ChangeBusinessType.vue:
```diff
@@ -133,26 +133,47 @@
-        </div>
+          <p class="pl-8">
+            A BC Community Contribution Company requires at least three directors.
+            File a director change and then come back and update the business type.
```
This seems OK to me but I thought we had decided to only warn users if they had less than the minimum but allow the alteration anyway (like we do if a CCC files a Change of Directors with less than 3 directors). Please confirm with Mihai that this is OK.

severinbeauvais replied:
If this is turned into a warning, change the colour to gold :)

severinbeauvais replied:
Jacqueline will discuss this with Janis on Monday. If Janis is not available then we will make a decision anyway.

severinbeauvais replied:
And if you are ready to merge today then go with your existing changes.

We can create another PR next week if needed.

severinbeauvais commented on src/components/common/YourCompany/ChangeBusinessType.vue:
```diff
@@ -133,26 +133,47 @@
+            >
+              mdi-alert
+            </v-icon>
+            <strong class="pl-2 pt-6">Update directors</strong>
```
I wonder if using [HTML Semantic Elements](https://www.w3schools.com/html/html5_semantic_elements.asp) would be a good idea here to improve accessibility? For example, this is definitely a header, and the body below is an article (or section).

severinbeauvais commented on src/components/common/YourCompany/CorrectName/CorrectName.vue:
```diff
@@ -157,6 +161,9 @@ export default class CorrectName extends Vue {
       this.formType = this.currentFormType
+      if (this.currentFormType === 'correct-name-to-number') {
+        this.setNameChangedToNumber(true)
+      }
```
Can this be done in the parent instead?

There's a [shared component](https://github.com/bcgov/bcrs-shared-components/tree/main/src/components/correct-name) for the components in this folder but it's slightly incompatible with the implementation here. There is [ticket 20827](https://app.zenhub.com/workspaces/entities---olga-65af15f59e89f5043c2911f7/issues/gh/bcgov/entity/20827) to use the shared component here so I don't want to make this code mode incompatible if possible.

severinbeauvais commented on src/components/common/YourCompany/EntityName.vue:
```diff
@@ -323,6 +325,25 @@ export default class EntityName extends Mixins(CommonMixin, NameRequestMixin) {
+      return 'LTD.'
+    }
+    return 'LTD.' // should never happen
+  }
```
This is almost a repeat of the method in CorrectNameToNumber. It would be nice to have just one method. Thoughts?

(This isn't worth a lot of time to fix. Is there an easy change?)

ArwenQin replied:
Moved the method to store.
For both components, can use param "this.getEntityType" to determine what to display

severinbeauvais commented on src/components/common/YourCompany/ChangeBusinessType.vue:
```diff
@@ -139,19 +139,23 @@
-            File a director change and then come back and update the business type.
-          </p>
+          <header>
+            <p>
```
You don't need the paragraph element. (I think header is equivalent to a div, but we mostly use p for styling and that's probably not needed here.) Test to make sure.

ArwenQin replied:
tried to remove <p>, but the font of the header is gonna be smaller:
![image](https://github.com/user-attachments/assets/c22082b7-54ee-4232-9bb3-2ec06d380963)

with <p>, same with design:
![image](https://github.com/user-attachments/assets/78470d69-9739-4f04-9cc6-ae588e4da379)

severinbeauvais replied:
Ah, so there's some style that a p has but a header doesn't.  You can add extra styles using the usual helper classes.

ArwenQin replied:
Thanks. I just updated the style, it looks same as Design now: (updated color to dark-grey, and font-14)
![image](https://github.com/user-attachments/assets/faabc2e3-fc98-4b86-851e-adf547593c21)

ArwenQin commented on src/components/common/YourCompany/EntityName.vue:
```diff
@@ -347,20 +350,23 @@ export default class EntityName extends Mixins(CommonMixin, NameRequestMixin) {
    * - the filing is an alteration or firm change filing, and
    * - the name has not been changed by type
+   * - this is when a named business is changed to a numbered business
    */
```
This method only returns true when changing a named business to numbered. So we can set the this.setNameChangedToNumber(result) here

severinbeauvais reviewed (approved):
Looks good. Make sure your tests still pass. If you don't get an answer from Janis tomorrow (on CCC min 3 dirs) then I'm OK with what you have here.

Oh, and please get another reviewer to approve :)

ArwenQin commented:
Thanks Severin, and I just fixed a bug:
This is a ULC, open the type change selection, select the same type, the error msg shows
![image](https://github.com/user-attachments/assets/b8d9b32f-eb1b-4945-8444-dc6921c4b7be)
I think it might be better to hide the error msg when select the same type

severinbeauvais commented:
> Thanks Severin, and I just fixed a bug: This is a ULC, open the type change selection, select the same type, the error msg shows !
> I think it might be better to hide the error msg when select the same type

This might a bug you introduced, because I don't remember an error when I changed from type X to type X previously.

Also, in Dev, I found another bug. Can you reproduce it? Open this alteration and change business type to ULC, then click the checkbox and click Done -- the company name is wrong (and should actually be unchanged).
https://dev.edit.business.bcregistry.gov.bc.ca/C9900853/alteration?accountid=3040

ArwenQin commented:
> Also, in Dev, I found another bug. Can you reproduce it? Open this alteration and change business type to ULC, then click the checkbox and click Done -- the company name is wrong (and should actually be unchanged).

Yes I see. I will fix it!
I noticed:
Numbered BC/ Named BC to Named/Numbered ULC - fee $1,000, is it expected?
However, if the user selects another type, gets the fee first, then click "Change":
![image](https://github.com/user-attachments/assets/c33cd37b-0372-44ad-ba1c-8141a86f6e01)

Then select ULC, the fee is not updated to $1,000:
![image](https://github.com/user-attachments/assets/de3f1ad8-695c-4ac2-b35b-9574bda9e73a)

We may need a new ticket for this bug

meawong commented:
> > Also, in Dev, I found another bug. Can you reproduce it? Open this alteration and change business type to ULC, then click the checkbox and click Done -- the company name is wrong (and should actually be unchanged).
> 
> Yes I see. I will fix it! I noticed: Numbered BC/ Named BC to Named/Numbered ULC - fee $1,000, is it expected? However, if the user selects another type, gets the fee first, then click "Change": ![image](https://private-user-images.githubusercontent.com/122495122/390926694-c33cd37b-0372-44ad-ba1c-8141a86f6e01.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI4NDQzMDMsIm5iZiI6MTczMjg0NDAwMywicGF0aCI6Ii8xMjI0OTUxMjIvMzkwOTI2Njk0LWMzM2NkMzdiLTAzNzItNDRhZC1iYTFjLTgxNDFhODZmNmUwMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyOVQwMTMzMjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lNDJhOGQ3MTQ4MmIzY2Q2NDg5YjMwYjkzODM0MGQyMmRmZjFjMjk1Zjk0MzdjNzUyNzQ3MGRiYzcwZTEyNThmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.9XCHkE-iNBF5LELilXOI55sfkvrrr3N12j2C_o9eN6k)
> 
> Then select ULC, the fee is not updated to $1,000: ![image](https://private-user-images.githubusercontent.com/122495122/390926781-de3f1ad8-695c-4ac2-b35b-9574bda9e73a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI4NDQzMDMsIm5iZiI6MTczMjg0NDAwMywicGF0aCI6Ii8xMjI0OTUxMjIvMzkwOTI2NzgxLWRlM2YxYWQ4LTY5NWMtNGFjMi1iMzViLTk1NzRiZGE5ZTczYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyOVQwMTMzMjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yZmM1YTFkYjkzZWIzNGIzNjFkMjNiYThlYTg0NDNkYjI0Nzg2MjRkYzE2OGMyYTUyNGRkMGY1MWNjMDlmZmUwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Rvh7px9lX4U76gKOL9NoS4QUDbbqag9WPdhWsfD0AMc)
> 
> and vice versa - select ULC, gets the $1,000 fee, the click "Change" and select another type, it's not updated to $100 We may need a new ticket for this bug Or, I think we can hide "Change" button here: ![image](https://private-user-images.githubusercontent.com/122495122/390927080-4585654f-0bc8-41ae-980e-baf14c58f14c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI4NDQzMDMsIm5iZiI6MTczMjg0NDAwMywicGF0aCI6Ii8xMjI0OTUxMjIvMzkwOTI3MDgwLTQ1ODU2NTRmLTBiYzgtNDFhZS05ODBlLWJhZjE0YzU4ZjE0Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyOVQwMTMzMjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xMDNhZjkxYmQyNTNjZWY1MjZjNWY2NGYzMWU2MDZhNWEwMzA2NDk1YjRiNTVkZTM0YjgzODMxZDY3OTQwNDUyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.b-h0x49_QDCIgSuyx-2o6HUdzio2PT-Deu22t_em6o4)



https://github.com/user-attachments/assets/68d4bc77-874d-40b0-bc1c-e1b58922bd70

I noticed the fee summary appears to update properly when you go from BC -> ULC -> BEN (latter part of the video) but not if you go from BC -> CCC -> ULC (first part of the video)? Also not captured in the video, but I tried a bunch of different subsequent alterations between all types but CCC (i.e. BC, BEN, ULC) and don't encounter an issue with the fee summary updating but i do see an issue when altering from CCC -> any other type (that isn't the original). Not sure if that insight is of any help.

meawong commented:
/gcbrun

meawong commented:
> Temporary Url for review: https://business-edit-dev--pr-600-w43kprmx.web.app

![image](https://github.com/user-attachments/assets/070aacde-f9eb-4e7a-a4f3-b575abda46cd)

bcregistry-sre commented:
Temporary Url for review: https://business-edit-dev--pr-600-w43kprmx.web.app

ArwenQin commented:
Thank you Megan. I agree with your findings. In summary:
1. The "Change" button of type change section: 
Yes, the bug only exists when change from/to CCC - Fee is not properly updated.

2. The "Change" button of name change section: 
Enter a CCC NR, get the fee
Then "Change" to a ULC NR, the fee is not updated. Vice versa.
In addition, enter a ULC NR first, then change to a CCC NR, the 3 directors error won't show.


I will fix these bugs in this ticket

ArwenQin commented:
> Also, in Dev, I found another bug. Can you reproduce it? Open this alteration and change business type to ULC, then click the checkbox and click Done -- the company name is wrong (and should actually be unchanged). https://dev.edit.business.bcregistry.gov.bc.ca/C9900853/alteration?accountid=3040

@severinbeauvais 
- I fixed the bug you mentioned - wrong ULC company name
- I fixed the bug in Type change section - when click "Change" button:
   fee amount didn't update when change from/to CCC ($100 ->$1,000)
- I fixed the bug in Name change section - "Change" button:
  enter a NR, click done, click done in type change section-> and then "Change" -> 
  enter a new NR -> the legal type section won't show the error msg and article confirmation check box

ArwenQin commented:
/gcbrun

bcregistry-sre commented:
Temporary Url for review: https://business-edit-dev--pr-600-w43kprmx.web.app

ArwenQin commented on src/components/common/YourCompany/ChangeBusinessType.vue:
```diff
@@ -348,6 +374,10 @@ export default class ChangeBusinessType extends Mixins(CommonMixin) {
     this.confirmArticles = false
+    // Ensure when selected type changes by NR, show the error message and Articles
+    if (this.selectedEntityType !== this.getOriginalLegalType) {
+      this.isEditingType = true
```
This is to fix the bug when click "Change" in name change section: enter a new NR that changes the business to a different type, the confirmArticles and error msg won't show.
The if condition is to avoid expanding this section when the component is initially loaded

severinbeauvais replied:
I love the comments explaining why this code exists.

ArwenQin commented on src/views/Alteration.vue:
```diff
@@ -216,6 +216,8 @@ export default class Alteration extends Mixins(CommonMixin, FeeMixin, FilingTemp
   @Watch('hasBusinessTypeChanged')
+  @Watch('getNameRequestLegalName')
+  @Watch('getEntityType')
   onBusinessTypeChanged () {
```
This is to fix the fee summary bug - 
sometimes the EntityType and BusinessName did change, but hasBusinessNameChanged/hasBusinessTypeChanged is not changed (keeps being true)
so add these 2 watchers.

ArwenQin commented:
/gcbrun

ArwenQin commented on src/views/Alteration.vue:
```diff
@@ -216,15 +216,20 @@ export default class Alteration extends Mixins(CommonMixin, FeeMixin, FilingTemp
+        this.getEntityType === CorpTypeCd.BC_ULC_COMPANY ||
+        this.getEntityType === CorpTypeCd.ULC_CONTINUE_IN
+      )
     ) {
```
Previously, C to ULC show $100. Now fixed by adding ULC here

severinbeauvais replied:
Good catch. Yes, old code was missing this. 👍

severinbeauvais commented on src/resources/Alteration/C.ts:
```diff
@@ -12,10 +12,10 @@ export const AlterationResourceC: ResourceIF = {
+  // Conditionally used in place of filingData when switching from or C to CUL:
   additionalFilingData: {
     filingTypeCode: FilingCodes.ALTERATION_BC_TO_ULC,
-    entityType: CorpTypeCd.BC_COMPANY,
```
Does this need correcting in the other resource files (CBEN, CCC, CUL)?

ArwenQin replied:
I believe no - only C/BC can alter to CUL/ULC

severinbeauvais replied:
Yes, you are right: there is no `additionalFilingData` for CBEN, CCC, CUL. 👍 

Please fix typo in comment ("from or C to CUL").

ArwenQin replied:
Thank you, fixed the typo

bcregistry-sre commented:
Temporary Url for review: https://business-edit-dev--pr-600-w43kprmx.web.app

severinbeauvais reviewed (approved):
I still approve, whenever you're ready.

Megan, does it look OK to you now?
